staging/src/k8s.io/apiserver/pkg/storage/cacher/delegator.go

```diff
@@ -219,7 +219,7 @@ func (c *CacheDelegator) GetList(ctx context.Context, key string, opts storage.L
 	success := "true"
 	fallback := "false"
 	if err != nil {
-		if errors.IsResourceExpired(err) {
+		if errors.IsResourceExpired(err) && utilfeature.DefaultFeatureGate.Enabled(features.ListFromCacheSnapshot) {
```

Comment:
This makes sense, the fallback here was designed for `ListFromCacheSnapshot` feature.

Reply:
it's a little hard to follow when these fallbacks (here and below in the `if result.ConsistentRead { if storage.IsTooLargeResourceVersion(err) {` block) should be looking at feature gates directly and when it should be looking at result.ConsistentRead / result.ShouldDelegate

it's also hard to follow which of these paths are for the `ListFromCacheSnapshot` feature and which are for the `ConsistentListFromCache` feature

any way we can clarify / explain this with comments or something to help our future selves?

Reply:
Added. LMK if its enough.